File: src/core/state.py
"""Interview state definition for LangGraph."""

from typing import Any, TypedDict


class InterviewState(TypedDict):
    """State for interview conversation flow.

    This defines the data structure that flows through the LangGraph.
    """

    # Session identifiers
    session_id: str
    user_id: str
    template_id: str

    # Conversation state
    current_topic_index: int
    conversation_history: list[dict[str, str]]
    extracted_info: dict[str, Any]

    # Topic/content
    topics: list[dict[str, Any]]
    topic: str  # Interview topic/title
    domain_context: str  # Domain knowledge context

    # Flow control
    status: str  # "planning", "interviewing", "analyzing", "completed"

    # Follow-up tracking (FR-002 AC-003: max 2 follow-ups per topic)
    followup_count: dict[str, int]  # topic_id -> count of follow-ups asked

    # Pending actions
    pending_question: str | None
    needs_followup: bool | None
    followup_type: str | None  # "clarification", "deep", "validation", "expansion"
    followup_reason: str | None  # Reason for follow-up

    # Report
    report: str | None
    report_path: str | None

    # Persona configuration
    persona_config: dict[str, Any] | None  # Role, personality, style settings


def create_initial_state(
    session_id: str,
    user_id: str,
    template_id: str = "quality_survey",
    topic: str = "访谈",
) -> InterviewState:
    """Create initial interview state.

    Args:
        session_id: Unique session identifier
        user_id: User identifier
        template_id: Interview template ID
        topic: Interview topic

    Returns:
        Initial InterviewState

    """
    return {
        "session_id": session_id,
        "user_id": user_id,
        "template_id": template_id,
        "current_topic_index": 0,
        "conversation_history": [],
        "extracted_info": {},
        "topics": [],
        "topic": topic,
        "domain_context": "",
        "status": "planning",
        "followup_count": {},  # FR-002 AC-003: Initialize empty follow-up count per topic
        "pending_question": None,
        "needs_followup": None,
        "followup_type": None,
        "followup_reason": None,
        "report": None,
        "report_path": None,
        "persona_config": None,
    }

File: src/core/state.ts
import { Annotation } from '@langchain/langgraph';
import type { InterviewTemplate, Message } from './types';

/**
 * LangGraph state annotation for type-safe state updates.
 * This replaces the plain TypeScript interface with LangGraph's Annotation system,
 * which provides proper reducer semantics for state updates.
 */
export const InterviewStateAnnotation = Annotation.Root({
  // Session identifiers
  sessionId: Annotation<string>,
  templateId: Annotation<string>,
  template: Annotation<InterviewTemplate>,

  // Conversation state
  conversationHistory: Annotation<Message[]>,
  currentTopicIndex: Annotation<number>,
  currentQuestionIndex: Annotation<number>,
  answers: Annotation<Record<string, string>>,
  completedTopics: Annotation<string[]>,

  // Flow control
  interviewStatus: Annotation<'planning' | 'interviewing' | 'followup' | 'analyzing' | 'completed'>,
  followupNeeded: Annotation<boolean>,
  followupQuestion: Annotation<string | undefined>,

  // Timing
  startTime: Annotation<Date>,
  endTime: Annotation<Date | undefined>,

  // Output
  report: Annotation<string | undefined>,
  error: Annotation<string | undefined>,
});

/**
 * InterviewState type derived from the annotation.
 * This type is used throughout the application for type safety.
 */
export type InterviewState = typeof InterviewStateAnnotation.State;

/**
 * Create initial state for a new interview session.
 * This is the entry point for starting a new interview.
 */
export function createInitialState(
  sessionId: string,
  templateId: string,
  template: InterviewTemplate
): InterviewState {
  return {
    sessionId,
    templateId,
    template,
    conversationHistory: [],
    currentTopicIndex: 0,
    currentQuestionIndex: 0,
    answers: {},
    completedTopics: [],
    interviewStatus: 'planning',
    followupNeeded: false,
    followupQuestion: undefined,
    startTime: new Date(),
    endTime: undefined,
    report: undefined,
    error: undefined,
  };
}

/**
 * Update conversation history with a new message.
 * Returns a partial state update that can be merged with the current state.
 */
export function updateConversationHistory(
  state: InterviewState,
  message: Message
): Partial<InterviewState> {
  return {
    conversationHistory: [...state.conversationHistory, message],
  };
}

/**
 * Extract the last user answer from conversation history.
 * Used by interview nodes to get the user's latest response.
 */
export function extractUserAnswer(state: InterviewState): string | null {
  const userMessages = state.conversationHistory.filter(
    (msg) => msg.role === 'user'
  );
  const lastUserMessage = userMessages[userMessages.length - 1];
  return lastUserMessage ? lastUserMessage.content : null;
}